README.md
# CDF-Quest


# Updates
Updates are usually pushed in 2-3 stages:



- Linux

Linux is pushed first, because that is what i use to primarily code and develop CDF-Quest. as such when it comes to pushing it requires no further testing.

- Windows

Windows comes second, because it does not require me to switch systems, as i can use wine to compile and test the windows version. This is also why sometimes the windows and linux versions are pushed together.

- macOS

macOS comes last, because i have to switch to a different system and pull the already pushed version, compile for mac os, then re push.

# ChangeLog
### Alpha Version 16

- Removed /theme command

- Reworked tile set

- Tiles have been updated and tile IDs have been updated.

- Added "interior shadow" tile attribute.

- Added tile information for 8 tiles

- Default world has been updated to showcase the new shadow effect

### Alpha Version 15

Added Version Designation to debug menu

- This is to establish what branch your version was pulled from on the github

Added Global Lighting

- This can be used to set a day night cycle, so that night has a dark environment

Added /lightlevel (/ll) command

- This is just used to set the light level via command


### Alpha Version 14

Fixed tile overlap bug

Fixed screen tearing bug

Added shadows (must be set manually by tile data for now)

added 4 new commands 
    /groundtile (/gt)
    /walltile (/wt)
    /ceilingtile (/ct)
    /tiledata   (/td)
    these 4 commands replace the /set-tile (/st) command.
    
    
Collision has been re added, however
    it is a tiledata flag and must be manually set for each tile for now
    
    
### Alpha Version 13

Complete Refactor
    The refactor in this update is not complete, as such certian functions are broken, future versions will finish this.
    
Player position is saved with the world, in the manifest file.

Added the ability to save a spawn point for when the player dies, also saved in manifest file

Old World are completely incompatable with A13 and up
    there will be an upgrade utility soon, when the new world format is finalized, to manually upgrade a map, create a new world under the name you wish to use (/new), close the game, put your *map* file in the maps folder of the new world, delete the " 0 0-0.cdf" file and rename the old map file to " 0 0-0.cdf" (the first space is important), the map should now load.

changed how tiles are set up
    instead of a background layer and a tile layer, there is now 3 tile layers, all able to be modified, and a tiled background texture to prevent black background from showing, there is also a 4th array that will have tile position properties such as colision and container id, updated upon tile update, saved seperate from tiles, also allowing for more easy tile adding
    
added a tile info file
    will have things such as what tile does what, what layer it can be placed on, and things like that
    
new command /new to create a new world
    

### Alpha Version 12

Updated Hub World
    Added path from door to well.

Worlds folder has been moved into assets
    This is just to clean up the root game directory
    
Error 101 is now part of error 100
    this change is due to the worlds folder being moved to part of the assets folder

Added world version checking
    checks to make sure the the world being loaded is compatable with the current version, if not error 103 is triggered, and allows you to update the manifest file of the world to match current version if you are sure that it is compatable. otherwise you can return to where you were, or quit to desktop
    
    
### Alpha Version 11

Minor Refactoring
    Refactored the code a bit, doesnt change much from gameplay perspective

Added Debug world
    Nothing really interesting, just a world that has all possible tiles, might make it autogenerated later, for now it is its own world

Added Gold and Experience values to character data type
    Currently does nothing yet, just a placeholder as is the level and points values.
    

### Alpha Version 10

Removed legacy map editor controls
    You can no longer hold "i" and press number keys to place tiles, this has been replaced by the new map editor mode
    
Added New Map Editor mode
    Map editor mode allows you to easily place tiles in the world use the number keys to place tiles, hold shift and press the number to cycle the tile it places, organized in 
        1: ground tiles
        2: decorative tiles
        3: container tiles
        4: roof tiles
        5: wall tiles
        6: door tiles
        
Added /update command
    this command allows you to update a map that you are editing without any text promps, this also does not touch the world manifest file. this function will also be used to save anything youve done in an area of the game when you leave, such as opened chests, killed entities, etc.

Removed Placeholder Titlescreen
    CDF-Quest will now default load "Hub" world. if this world does not exist, it will throw an error and, if you continue, will take you to a blank world with default values. use /gamemode and choose 1 to enter map editor mode
    
    
### Alpha Version 9

Added Protected World password
    This prevents worlds from being accidentally overwriten if you do not provide the protection password that is matched in the manifest file, this is not a security feature and is easily bypassed, it exists soley to prevent saving to the wrong world on accident.

Added file extention to world files
    There was no issue with the game and no extentions, however when viewed in a file browser, the extentionless files could act strange, so ive added the cdf extention to both the manifest file, and the map files, this extention is contextual, as what is in the file and what it does is based on where its at in the folder structure and its title.
 


A9/Changelog
Version Alpha 9

Added Protected World password
    This prevents worlds from being accidentally overwriten if you do not provide the protection password that is matched in the manifest file, this is not a security feature and is easily bypassed, it exists soley to prevent saving to the wrong world on accident.

Added file extention to world files
    There was no issue with the game and no extentions, however when viewed in a file browser, the extentionless files could act strange, so ive added the cdf extention to both the manifest file, and the map files, this extention is contextual, as what is in the file and what it does is based on where its at in the folder structure and its title.


CDF-Quest.bas
$NoPrefix
Option Explicit
On Error GoTo ERRORHANDLER
Randomize Timer
Screen NewImage(641, 481, 32) '40x30
PrintMode KeepBackground
Title "CDF-Quest"


'$include: 'Assets\Sources\VariableDeclaration.bi'

'$include: 'Assets\Sources\DefaultValues.bi'

'$include: 'Assets\Sources\TileIndex.bi'

'$include: 'Assets\Sources\InventoryIndex.bi'


INITIALIZE


GoTo game
Error 102
ERRORHANDLER: ERRORHANDLER
game:

Do
    SETBG
    SetMap
    CastShadow
    MOVE
    COLDET
    SPSET
    INTER
    ZOOM
    SetLighting
    HUD
    If Flag.InventoryOpen = 1 Then InventoryUI
    UseItem
    DEV
    KeyPressed = KeyHit
    If Flag.FrameRateLock = 0 Then Limit Settings.FrameRate
    CurrentTick = CurrentTick + Settings.TickRate
    If Flag.ScreenRefreshSkip = 0 Then Display
    Flag.ScreenRefreshSkip = 0
    Cls
Loop



Error 102

Sub InventoryUI

End Sub

Sub UseItem

End Sub

Sub INTER
    Select Case KeyPressed
        Case -15616
            Flag.DebugMode = Flag.DebugMode + 1
        Case -15104
            Flag.HudDisplay = Flag.HudDisplay + 1
        Case -101
            Flag.InventoryOpen = Flag.InventoryOpen + 1


    End Select
End Sub



Sub SetLighting
    Dim i As Byte
    Dim ii As Byte
    For i = 0 To 30
        For ii = 0 To 40
            PutImage (ii * 16, i * 16)-((ii * 16) + 15.75, (i * 16) + 15.75), Texture.Shadows, , (16 * (GlobalLightLevel - LocalLightLevel(ii, i)), 16)-(16 * (GlobalLightLevel - LocalLightLevel(ii, i)) + 15, 31)
        Next
    Next
End Sub




Sub HUD
    If Flag.HudDisplay = 0 Then
        Dim tmpheal As Byte
        Dim token As Byte
        Dim hboffset As Byte
        Dim hbpos As Byte
        Dim hbitemsize As Single
        Dim invrow As Byte
        Dim invoffset As Byte
        Dim invheight As Byte
        invoffset = 1
        invheight = 5
        hboffset = 1
        token = 1
        hbitemsize = 2

        tmpheal = Player.health

        While tmpheal > 0

            If tmpheal <= 8 Then
                PutImage (CameraPositionX + 88 - 16, CameraPositionY - 52 + (token - 1) * 16)-(CameraPositionX + 88, CameraPositionY - 52 + 16 + (token - 1) * 16), Texture.HudSprites, , ((tmpheal - 1) * 32, 0)-((tmpheal - 1) * 32 + 31, 31)
            Else
                PutImage (CameraPositionX + 88 - 16, CameraPositionY - 52 + (token - 1) * 16)-(CameraPositionX + 88, CameraPositionY - 52 + 16 + (token - 1) * 16), Texture.HudSprites, , (7 * 32, 0)-(7 * 32 + 31, 31)
            End If
            tmpheal = tmpheal - 8
            token = token + 1
        Wend

        'Hotbar
        For hbpos = 0 To 5
            PutImage (CameraPositionX - 72 + hboffset + (17 * hbpos), CameraPositionY + 68 - 16 - hboffset)-(CameraPositionX - 72 + 16 + hboffset + (17 * hbpos), CameraPositionY + 68 - hboffset), Texture.HudSprites, , (0, 32)-(31, 63)
            PutImage (CameraPositionX - 72 + hboffset + (17 * hbpos) + hbitemsize, CameraPositionY + 68 - 16 - hboffset + hbitemsize)-(CameraPositionX - 72 + 16 + hboffset + (17 * hbpos) - hbitemsize, CameraPositionY + 68 - hboffset - hbitemsize), Texture.ItemSheet, , (Inventory(1, hbpos + 1, 1), Inventory(1, hbpos + 1, 2))-(Inventory(1, hbpos + 1, 1) + 15, Inventory(1, hbpos + 1, 2) + 15)
        Next


        If Flag.InventoryOpen = 1 Then

            Select Case GameMode
                Case 1
                    For invrow = 0 To 2
                        For hbpos = 0 To 5
                            PutImage (CameraPositionX - 72 + hboffset + (17 * hbpos), (CameraPositionY + 68 - 16 - hboffset) - (16 * (invrow + 1) + invoffset * invrow) - invheight)-(CameraPositionX - 72 + 16 + hboffset + (17 * hbpos), (CameraPositionY + 68 - hboffset) - (16 * (invrow + 1) + invoffset * invrow) - invheight), Texture.HudSprites, , (0, 32)-(31, 63)
                        Next
                    Next

            End Select

        End If

    End If


    '     Case 0
    '        For hbpos = 0 To 5
    '           PutImage (CameraPositionX - 72 + hboffset + (17 * hbpos) + hbitemsize, CameraPositionY + 68 - 16 - hboffset + hbitemsize)-(CameraPositionX - 72 + 16 + hboffset + (17 * hbpos) - hbitemsize, CameraPositionY + 68 - hboffset - hbitemsize), Texture.TileSheet, , (Inventory(1, hbpos + 1) * 16, hbpos * 16)-(Inventory(1, hbpos + 1) * 16 + 15, hbpos * 16 + 15)
    '      Next
    '
    '
    '
    ' If KeyDown(49) Then GroundTile(Int((Player.x + 8) \ 16) + 1, Int((Player.y + 8) \ 16) + 1) = Inventory(1, 1)

    '   If KeyDown(50) Then GroundTile(Int((Player.x + 8) \ 16) + 1, Int((Player.y + 8) \ 16) + 1) = Inventory(1, 2) + 16
    '
    '   If KeyDown(51) Then GroundTile(Int((Player.x + 8) \ 16) + 1, Int((Player.y + 8) \ 16) + 1) = Inventory(1, 3) + 32
    '
    '   If KeyDown(52) Then GroundTile(Int((Player.x + 8) \ 16) + 1, Int((Player.y + 8) \ 16) + 1) = Inventory(1, 4) + 48
    '
    '   If KeyDown(53) Then GroundTile(Int((Player.x + 8) \ 16) + 1, Int((Player.y + 8) \ 16) + 1) = Inventory(1, 5) + 64
    '
    '   If KeyDown(54) Then GroundTile(Int((Player.x + 8) \ 16) + 1, Int((Player.y + 8) \ 16) + 1) = Inventory(1, 6) + 80
    '
    '   Select Case KeyPressed
    '       Case 33
    '           Inventory(1, 1) = Inventory(1, 1) + 1
    '           If Inventory(1, 1) > 5 Then Inventory(1, 1) = 0
    '       Case 64
    '           Inventory(1, 2) = Inventory(1, 2) + 1
    '           If Inventory(1, 2) > 4 Then Inventory(1, 2) = 0
    '       Case 35
    '           Inventory(1, 3) = Inventory(1, 3) + 1
    '           If Inventory(1, 3) > 3 Then Inventory(1, 3) = 0
    '       Case 36
    '           Inventory(1, 4) = Inventory(1, 4) + 1
    '           If Inventory(1, 4) > 9 Then Inventory(1, 4) = 0
    '       Case 37
    '           Inventory(1, 5) = Inventory(1, 5) + 1
    '           If Inventory(1, 5) > 4 Then Inventory(1, 5) = 0
    '       Case 94
    '           Inventory(1, 6) = Inventory(1, 6) + 1
    '           If Inventory(1, 6) > 1 Then Inventory(1, 6) = 0
    '
    '        End Select


End Sub




Sub DEV
    If Flag.DebugMode = 1 Then
        PrintMode FillBackground
        Dim comin As String
        Dim dv As Single
        Dim dummystring As String
        Dim databit As Byte



        Locate 1, 1
        ENDPRINT "Debug Menu (Press F3 to Close)"
        Print
        ENDPRINT "Version: " + Game.Version
        ENDPRINT "Version Designation: " + Game.Designation
        ENDPRINT "Operating System: " + Game.HostOS
        If Game.32Bit = 1 Then ENDPRINT "32-Bit Compatability Mode"
        Print
        ENDPRINT "Flags:"
        If Flag.StillCam = 1 Then ENDPRINT "Still Camera Enabled"
        If Flag.FreeCam = 1 Then ENDPRINT "Free Camera Enabled"
        If Flag.FullCam = 1 Then ENDPRINT "Full Camera Enabled"
        If Flag.NoClip = 1 Then ENDPRINT "No Clip Enabled"
        If bgdraw = 1 Then ENDPRINT "Background Drawing Disabled"
        If Flag.InventoryOpen = 1 Then ENDPRINT "Inventory Open"


        Locate 1, 1
        Print Game.Title; " ("; Game.Buildinfo; ")"
        Print
        Print "FPS:" + Str$(FRAMEPS) + " /" + Str$(CurrentTick)
        Print "Window:"; CameraPositionX; ","; CameraPositionY
        Print "Current World: "; WorldName; " (" + SavedMap + ")"
        Print "Light Level: (G:"; GlobalLightLevel; ", L: 0)"
        Print "Gamemode: ";
        Select Case GameMode
            Case 0
                Print "Legacy Map Editor"
            Case 1
                Print "Creative"
            Case 2
                Print "Survival"
            Case 3
                Print "Combat"
            Case 4
                Print "Hub"
        End Select

        Print
        Print "Data Viewer: ";
        Select Case Debug.Tracking
            Case ""
                Print "None Selected"
                Print "Start tracking an entity to view its data"
            Case "player", "1"
                Print "Player"
                Print "POS:"; Player.x; ","; Player.y; "("; Int((Player.x + 8) / 16) + 1; ","; Int((Player.y + 8) / 16) + 1; ")"
                Print "Facing:"; Player.facing
                Print "Motion:"; Player.moving
                Print "Contacted Tile ID:"; Player.tile; "(" + Hex$(Player.tile) + ")"
                Print "Facing Tile ID:"; Player.tilefacing; "(" + Hex$(Player.tilefacing) + ")"
            Case Else
                Print "Unrecognized Tile or Entity"
        End Select

        If KeyDown(47) Then
            KeyClear

            Locate 28, 1: Input "/", comin
            Select Case comin
                Case "teleport", "tp"
                    Locate 28, 1: Print "               "
                    Locate 28, 1: Input "Teleport x: ", Player.x
                    Locate 28, 1: Print "               "
                    Locate 28, 1: Input "Teleport y: ", Player.y
                Case "tileport", "tip"
                    Locate 28, 1: Print "               "
                    Locate 28, 1: Input "Teleport x: ", dv
                    Player.x = dv * 16
                    Locate 28, 1: Print "               "
                    Locate 28, 1: Input "Teleport y: ", dv
                    Player.y = dv * 16

                Case "stillcam", "sc"
                    Flag.StillCam = Flag.StillCam + 1
                Case "fullcam", "fc"
                    Flag.FullCam = Flag.FullCam + 1
                Case "freecam", "frc"
                    Flag.FreeCam = Flag.FreeCam + 1
                Case "noclip", "nc"
                    Flag.NoClip = Flag.NoClip + 1
                Case "exit"
                    System
                Case "error"
                    Locate 28, 1: Print "                              "
                    Locate 28, 1: Input "Simulate error number: ", dv
                    Error dv
                Case "gamemode", "gm"
                    Locate 28, 1: Print "                         "
                    Locate 28, 1: Input "Change Gamemode to: ", GameMode
                Case "health"
                    Locate 28, 1: Print "                      "
                    Locate 28, 1: Input "Set Health to: ", Player.health
                Case "track", "tr"
                    Locate 28, 1: Print "                               "
                    Locate 28, 1: Input "Track Entity ID or Tile ID: ", Debug.Tracking
                Case "framerate-unlock", "fru"
                    Flag.FrameRateLock = Flag.FrameRateLock + 1
                Case "save"
                    SAVEMAP (0)
                Case "save1"
                    SAVEMAP (1)
                Case "load"
                    Locate 28, 1: Print "                                "
                    Locate 28, 1: Input "Name of Map File to load: ", map.filename
                    LOADMAP (map.filename)
                Case "loadworld"
                    Locate 28, 1: Print "                                "
                    Locate 28, 1: Input "Name of World Folder to load: ", WorldName
                    LOADWORLD
                Case "groundtile", "gt"
                    Locate 28, 1: Print "                   "
                    Locate 28, 1: Input "Set GroundTile ID: ", GroundTile(Int((Player.x + 8) / 16) + 1, Int((Player.y + 8) / 16) + 1)
                Case "walltile", "wt"
                    Locate 28, 1: Print "                 "
                    Locate 28, 1: Input "Set WallTile ID: ", WallTile(Int((Player.x + 8) / 16) + 1, Int((Player.y + 8) / 16) + 1)
                Case "ceilingtile", "ct"
                    Locate 28, 1: Print "                   "
                    Locate 28, 1: Input "Set CeilingTile ID: ", CeilingTile(Int((Player.x + 8) / 16) + 1, Int((Player.y + 8) / 16) + 1)
                Case "tiledata", "td"
                    Locate 28, 1: Print "                 "
                    Locate 28, 1: Input "Select Data Bit: ", databit
                    Locate 28, 1: Print "                   "
                    Locate 28, 1: Input "Select Data Value: ", TileData(Int((Player.x + 8) / 16) + 1, Int((Player.y + 8) / 16) + 1, databit)


                Case "bgdraw"
                    bgdraw = bgdraw + 1
                Case "update", "up"
                    UPDATEMAP

                Case "new"
                    NewWorld

                Case "lightlevel", "ll"
                    Locate 28, 1: Print "                    "
                    Locate 28, 1: Input "Select Light Level  ", GlobalLightLevel

                Case Else
            End Select
            Flag.ScreenRefreshSkip = 1

        End If
        PrintMode KeepBackground
    End If
End Sub


Sub SetMap
    Dim i As Byte
    Dim ii As Byte
    Dim tileposx As Integer
    Dim tileposy As Integer
    For i = 1 To 30
        For ii = 1 To 40
            tileposx = GroundTile(ii, i) * 16
            tileposy = 0
            While tileposx >= 256
                tileposx = tileposx - 256
                tileposy = tileposy + 16
            Wend
            PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.TileSheet, , (TileIndex(GroundTile(ii, i), 1), TileIndex(GroundTile(ii, i), 2))-(TileIndex(GroundTile(ii, i), 1) + 15, TileIndex(GroundTile(ii, i), 2) + 15)


            tileposx = WallTile(ii, i) * 16
            tileposy = 0
            While tileposx >= 256
                tileposx = tileposx - 256
                tileposy = tileposy + 16
            Wend
            PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.TileSheet, , (TileIndex(WallTile(ii, i), 1), TileIndex(WallTile(ii, i), 2))-(TileIndex(WallTile(ii, i), 1) + 15, TileIndex(WallTile(ii, i), 2) + 15)


            tileposx = CeilingTile(ii, i) * 16
            tileposy = 0
            While tileposx >= 256
                tileposx = tileposx - 256
                tileposy = tileposy + 16
            Wend
            PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.TileSheet, , (TileIndex(CeilingTile(ii, i), 1), TileIndex(CeilingTile(ii, i), 2))-(TileIndex(CeilingTile(ii, i), 1) + 15, TileIndex(CeilingTile(ii, i), 2) + 15)


        Next
    Next
End Sub

Sub CastShadow
    Dim i, ii As Byte
    For i = 1 To 30
        For ii = 1 To 40

            If TileData(ii, i + 1, 1) = 1 And TileData(ii, i, 2) = 0 Then
                PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.Shadows, , (0, 0)-(15, 15)
            End If

            If TileData(ii + 1, i, 1) = 1 And TileData(ii, i, 2) = 0 Then
                PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.Shadows, , (16, 0)-(31, 15)
            End If

            If TileData(ii - 1, i, 1) = 1 And TileData(ii, i, 2) = 0 Then
                PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.Shadows, , (32, 0)-(47, 15)
            End If

            If TileData(ii, i - 1, 1) = 1 And TileData(ii, i, 2) = 0 Then
                PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.Shadows, , (48, 0)-(63, 15)
            End If

            If TileData(ii, i, 3) = 1 Then

                If TileData(ii, i + 1, 3) = 0 Then
                    PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.Shadows, , (0, 0)-(15, 15)
                End If

                If TileData(ii + 1, i, 3) = 0 Then
                    PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.Shadows, , (16, 0)-(31, 15)
                End If

                If TileData(ii - 1, i, 3) = 0 Then
                    PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.Shadows, , (32, 0)-(47, 15)
                End If

                If TileData(ii, i - 1, 3) = 0 Then
                    PutImage ((ii - 1) * 16, (i - 1) * 16)-(((ii - 1) * 16) + 15.75, ((i - 1) * 16) + 15.75), Texture.Shadows, , (48, 0)-(63, 15)
                End If

            End If

        Next
    Next
End Sub



Sub INITIALIZE

    If DirExists("Assets") Then
        If DirExists("Assets\Sprites") = 0 Then Error 100
        If DirExists("Assets\Sprites\Entities") = 0 Then Error 100
        If DirExists("Assets\Sprites\Items") = 0 Then Error 100
        If DirExists("Assets\Sprites\Other") = 0 Then Error 100
        If DirExists("Assets\Sprites\Tiles") = 0 Then Error 100
        If DirExists("Assets\Music") = 0 Then Error 100
        If DirExists("Assets\Sounds") = 0 Then Error 100
        If DirExists("Assets\Worlds") = 0 Then Error 100
        If DirExists("Assets\SaveData") = 0 Then MkDir "Assets\SaveData": new = 1
    Else Error 100
    End If



    If new = 1 Then SAVESETTINGS
    LOADSETTINGS

    OSPROBE

    Texture.PlayerSprites = LoadImage(File.PlayerSprites)
    Texture.TileSheet = LoadImage(File.TileSheet)
    Texture.ItemSheet = LoadImage(File.ItemSheet)
    Texture.HudSprites = LoadImage(File.HudSprites)
    Texture.Shadows = LoadImage(File.Shadows)




    WorldName = "Hub"
    LOADWORLD


End Sub


Sub ERRORHANDLER
    AutoDisplay
    Cls
    PLAYSOUND Sounds.error
    Delay 0.5
    KeyClear
    Locate 1, 1
    CENTERPRINT "CDF ERROR HANDLER"
    Print "Error Code:"; Err
    Locate 2, 1
    ENDPRINT "Error Line:" + Str$(ErrorLine)
    Print "--------------------------------------------------------------------------------"
    Print
    '       PRINT "--------------------------------------------------------------------------------"
    Select Case Err
        Case 100
            Print "Assets folder is incomplete, this error can be triggered by one or more of the"
            Print "following conditions:"
            Print
            Print
            Print "     The assets folder is missing"
            Print
            Print "     Sub-directories in the Assets folder are missing"
            Print
            Print "     The contents of assets, or the directory itself is corrupted"
            Print
            Print "     You do not have proper permissions to access the assets directory"
            Print
            Print
            Print "Make sure the entireity of the assets folder is present and accessible to your"
            Print "user account and, if necessary, redownload the assets folder."
            Print
            Print "The assets folder, and its contents are necessary for the game to load, as it"
            Print "contains all sprite and texture files, sounds and music, user saved data, and"
            Print "world files. Without these, the game will not play correctly. It is advised to"
            Print "not continue."
            CONTPROMPT

        Case 101
            Print "This is a legacy error code, and should never be triggered in game, if it has"
            Print "been triggered, not due to the /error command, please contact the developer"
            CONTPROMPT

        Case 102
            Print "Invalid Code Position, This error occurs when the program flow enters an area"
            Print "that it should not be, This is most likely a programming issue, and not an end"
            Print "user issue."
            Print ""
            Print "There is no user solution to this issue, please file a bug report to the"
            Print "developers, including the line number and what you were doing when it occured."
            CONTPROMPT

        Case 103
            Print "This world was not made for this version of "; Game.Title; ". This means one of"
            Print "the following cases is true:"
            Print
            Print
            Print "     You are attempting to load an out of date world"
            Print
            Print "     You are attempting to load a world designed for a newer version of"
            Print "     "; Game.Title
            Print
            Print "     Your world manifest is corrupted"
            Print
            Print
            Print "Double check the world version and game version."
            Print "World: ("; mapversion; ") Game: ("; Game.Version; ")"
            Print
            Print "If you are certain that this is a mistake, you may try to update the manifest"
            Print "here. Note that this does not update old worlds, just broken manifest files"
            Print "Otherwise you can try to load a different world. "; Game.Title; ""
            Print "does not support loading out of version worlds."
            Print
            Print
            CENTERPRINT "(U)pdate manifest, (R)eturn to existing map, (Q)uit to desktop."
            Do
                If KeyDown(113) Then System
                If KeyDown(114) Then Exit Do
                If KeyDown(117) Then
                    Open "Assets\Worlds\" + WorldName + "\Manifest.cdf" As #1
                    Put #1, 3, Game.Version
                    Close #1

                    Exit Do

                End If
            Loop


        Case Else
            Print "Unrecognized error, contact developers"
            CONTPROMPT

    End Select

    KeyClear
    Cls
    Resume Next
End Sub

Sub CONTPROMPT
    Print
    Print

    CENTERPRINT "(I)gnore this error and continue anyway, (Q)uit to desktop"
    Do
        If KeyDown(113) Then System
        If KeyDown(105) Then Exit Do
    Loop
End Sub





Sub SETBG
    If bgdraw = 0 Then
        Dim i As Byte
        Dim ii As Byte
        For i = 0 To 30
            For ii = 0 To 40
                PutImage (ii * 16, i * 16)-((ii * 16) + 15.75, (i * 16) + 15.75), Texture.TileSheet, , (16, 0)-(31, 15)
            Next
        Next
    End If
End Sub







Sub NewWorld

    Cls
    KeyClear
    AutoDisplay
    Input "World Name?", WorldName
    SavedMapX = 0
    SavedMapY = 0
    Player.x = 320
    Player.y = 200
    SAVEMAP (0)
    LOADWORLD

End Sub


'$include: 'Assets\Sources\Initialization.bm'
'$include: 'Assets\Sources\FileAccess.bm'
'$include: 'Assets\Sources\TextControl.bm'
'$include: 'Assets\Sources\ErrorHandler.bm'
'$include: 'Assets\Sources\FrameRate.bm'
'$include: 'Assets\Sources\OsProbe.bm'
'$include: 'Assets\Sources\CollisionDetection.bm'
'$include: 'Assets\Sources\SpriteAnimation.bm'
'$include: 'Assets\Sources\PlayerControl.bm'
'$include: 'Assets\Sources\MapDraw.bm'
'$include: 'Assets\Sources\ScreenZoom.bm'
'$include: 'Assets\Sources\AudioControl.bm'



